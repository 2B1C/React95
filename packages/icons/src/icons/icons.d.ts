interface ICO extends ArrayBuffer {}

declare module '*.ico' {
  const value: ICO;
  export = value;
}
