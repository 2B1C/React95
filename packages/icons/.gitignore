src/react/
src/index.ts
src/iconDemos.ts
png/
icons.css
