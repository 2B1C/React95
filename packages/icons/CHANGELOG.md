# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.4.2](https://github.com/React95/React95/compare/@react95/icons@0.4.1...@react95/icons@0.4.2) (2020-10-03)


### Bug Fixes

* **icons:** remove dictories and file fields from package.json and fix publishConfig ([2d788cc](https://github.com/React95/React95/commit/2d788cccaf41517e52cffc417731cba4f7e97087))





## [0.4.1](https://github.com/React95/React95/compare/@react95/icons@0.4.0...@react95/icons@0.4.1) (2020-10-03)


### Bug Fixes

* **icons:** add prepublish npm script ([cca8196](https://github.com/React95/React95/commit/cca8196129007bb5b33ba600cec5dd6bf13c678a))





# [0.4.0](https://github.com/React95/React95/compare/@react95/icons@0.3.0...@react95/icons@0.4.0) (2020-10-03)


### Features

* **icons:** add icons declaration file ([893b3ef](https://github.com/React95/React95/commit/893b3ef7f2fd639cdc1766907e4f62608bbfd8a3))
* **types:** move *.ico module declaration to root types folder ([bef505d](https://github.com/React95/React95/commit/bef505d89fcdfcbb1da7bad855f9eb03308191ba))





# [0.3.0](https://github.com/React95/React95/compare/@react95/icons@0.2.0...@react95/icons@0.3.0) (2020-06-14)


### Features

* **icons:** add new icons! :tada: ([feaf970](https://github.com/React95/React95/commit/feaf97094bdb7da37fae82d14871286e85e74ab4)), closes [#154](https://github.com/React95/React95/issues/154)





# [0.2.0](https://github.com/React95/React95/compare/@react95/icons@0.1.0...@react95/icons@0.2.0) (2020-05-15)


### Features

* **icons:** a lot of new icons! ([ceb0761](https://github.com/React95/React95/commit/ceb07614de289fba86e837dacf7399313e4a1460))





# 0.1.0 (2019-09-15)


### Features

* **Icon:** Add icons package ([94ec2af](https://github.com/React95/React95/commit/94ec2af)), closes [#124](https://github.com/React95/React95/issues/124)
