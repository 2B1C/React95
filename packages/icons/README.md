# `@react95/icons`

> TODO: description

## Usage

```
const icons = require('@react95/icons');

// TODO: DEMONSTRATE API
```
