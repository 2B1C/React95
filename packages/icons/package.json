{
  "name": "@react95/icons",
  "version": "1.0.3",
  "description": "React95 icons gallery",
  "keywords": [
    "React95",
    "icons",
    "windows95",
    "w95"
  ],
  "files": [
    "icons.css",
    "dist",
    "png"
  ],
  "author": "ggdaltoso <ggdaltoso@gmail.com>",
  "homepage": "https://react95.github.io/React95/",
  "license": "MIT",
  "main": "./src/index.ts",
  "types": "./dist/types/",
  "unpkg": "icons.css",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/React95/React95.git"
  },
  "bugs": {
    "url": "https://github.com/React95/React95/issues"
  },
  "scripts": {
    "precopy": "mkdir -p dist/src/png",
    "postinstall": "ts-node processIcons.ts",
    "build": "yarn build:cjs && yarn build:esm && yarn build:types",
    "build:types": "cross-env NODE_ENV=production tsc -p ./tsconfig.json",
    "build:esm": "cross-env NODE_ENV=esm yarn babel ./src --out-dir dist/esm --extensions .ts,.tsx",
    "build:cjs": "cross-env NODE_ENV=cjs yarn babel ./src --out-dir dist/cjs --extensions .ts,.tsx",
    "prepublish": "yarn build",
    "prepublishOnly": "node ../../scripts/prepublish.js"
  },
  "devDependencies": {
    "@types/rimraf": "^3.0.0",
    "rimraf": "^3.0.2",
    "ts-node": "^9.1.1",
    "typescript": "^4.2.3"
  }
}
