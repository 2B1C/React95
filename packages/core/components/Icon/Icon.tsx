import styled from 'styled-components';
import icons from '@react95/icons/icons.module.css';

export interface IconProps {
  name: keyof typeof icons;
}

const Icon = styled.i
  .withConfig({
    shouldForwardProp: (prop, defaultValidatorFn) =>
      !['name'].includes(prop) && defaultValidatorFn(prop),
  })
  .attrs<IconProps>(({ name }) => ({
    className: name,
  }))<IconProps>`
  display: block;
`;

export default Icon;
