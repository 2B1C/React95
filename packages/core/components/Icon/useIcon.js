import { useState, useEffect } from 'react';
import icons from '@react95/icons';
import ico from 'icojs';

const MIME_TYPE = 'image/png';

function icoParse(file) {
  if (ico.isICO(file)) {
    return ico.parse(file);
  }

  return [];
}

const useIcon = ({ name, size, fallback }) => {
  const [data, setAvailableIcons] = useState({
    iconUrl: '',
    availableIcons: [{}],
  });

  const changeIconUrl = newUrl => {
    setAvailableIcons(iconData => ({ ...iconData, iconUrl: newUrl }));
  };

  useEffect(() => {
    async function fetchIcon() {
      const response = await fetch(icons[name]);
      const iconBuffer = await response.arrayBuffer();

      const allIcons = await icoParse(iconBuffer);

      const iconsToRender = allIcons.map(i => ({
        size: i.width,
        url: URL.createObjectURL(new Blob([i.buffer], { type: MIME_TYPE })),
        bit: i.bbt,
      }));

      let url;
      if (fallback) {
        url = iconsToRender[0].url;
      } else {
        const match = iconsToRender.find(i => i.size === size);
        url = match ? match.url : undefined;
      }

      setAvailableIcons({ iconUrl: url, availableIcons: iconsToRender });
    }

    fetchIcon();
  }, []);

  return {
    changeIconUrl,
    iconUrl: data.iconUrl,
    availableIcons: data.availableIcons,
  };
};

export default useIcon;
