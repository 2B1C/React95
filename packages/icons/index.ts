import icons from './src';

export default icons;
