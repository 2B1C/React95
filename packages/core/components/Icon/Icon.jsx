import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import useIcon from './useIcon';

const I = styled.i`
  display: block;

  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;

  ${({ width, height, url }) => css`
    width: ${width}px;
    height: ${height}px;
    background-image: ${url ? `url('${url}')` : 'none'};
  `}
`;

const Icon = ({ name, width, height, size, fallback, ...rest }) => {
  const { iconUrl, changeIconUrl, availableIcons } = useIcon({
    name,
    size,
    fallback,
  });

  useEffect(() => {
    const icon = availableIcons.find(i => i.size === size);

    changeIconUrl(icon ? icon.url : undefined);
  }, [size]);

  return (
    <I
      name={name}
      width={width || size}
      height={height || size}
      {...rest}
      url={iconUrl}
    />
  );
};

Icon.propTypes = {
  name: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  fallback: PropTypes.bool,
  size: PropTypes.number,
};

Icon.defaultProps = {
  name: null,
  width: undefined,
  height: undefined,
  fallback: true,
  size: 32,
};

export default Icon;
