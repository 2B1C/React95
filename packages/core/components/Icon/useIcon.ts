import * as React from 'react';
import icons from '@react95/icons';
import ico from 'icojs';

import 'isomorphic-fetch';

import { IconProps } from './Icon';

const MIME_TYPE = 'image/png';

async function icoParse(file: ArrayBuffer): Promise<ico.ParsedImage[]> {
  if (ico.isICO(file)) {
    return ico.parse(file);
  }

  return [];
}

export interface IAvailableIcon {
  size: number;
  url: string;
  bpp: number;
}

type UseIconState = {
  iconUrl?: string;
  availableIcons: Array<IAvailableIcon>;
};

const useIcon = ({
  name,
  size,
  fallback,
  bpp = 4,
  variant = 1,
}: Partial<IconProps>) => {
  const [data, setAvailableIcons] = React.useState<UseIconState>({
    iconUrl: '',
    availableIcons: [],
  });

  const changeIconUrl = (newUrl: string) => {
    setAvailableIcons(iconData => ({ ...iconData, iconUrl: newUrl }));
  };

  React.useEffect(() => {
    async function fetchIcon() {
      const response = await fetch(icons[name]);
      const iconBuffer = await response.arrayBuffer();

      const allIcons = await icoParse(iconBuffer);

      const iconsToRender = allIcons.map(i => ({
        size: i.width,
        url: URL.createObjectURL(new Blob([i.buffer], { type: MIME_TYPE })),
        bpp: i.bpp,
      }));

      let url = undefined;
      const match = iconsToRender.filter(i => i.size === size && i.bpp === bpp);

      if (match.length > 0) {
        const found = match[variant - 1];
        url = found.url;
      } else if (fallback) {
        url = iconsToRender[0].url;
      }

      setAvailableIcons({ iconUrl: url, availableIcons: iconsToRender });
    }

    fetchIcon();
  }, []);

  return {
    changeIconUrl,
    iconUrl: data.iconUrl,
    availableIcons: data.availableIcons,
  };
};

export default useIcon;
