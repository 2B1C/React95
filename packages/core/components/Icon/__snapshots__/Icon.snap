// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should match snapshot 1`] = `
.c0 {
  display: block;
}

<div>
  <i
    class="c0 r95 bat_16x16_4bit"
  />
</div>
`;
